controller/cache/cluster.go

```diff
@@ -0,0 +1,228 @@
+package cache
+
+import (
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/argoproj/argo-cd/common"
+	appv1 "github.com/argoproj/argo-cd/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/util"
+	"github.com/argoproj/argo-cd/util/kube"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+const (
+	clusterSyncTimeout = 1 * time.Hour
+)
+
+type clusterInfo struct {
+	nodes        map[string]*node
+	lock         *sync.Mutex
+	onAppUpdated func(appName string)
+	kubectl      kube.Kubectl
+	cluster      *appv1.Cluster
+	syncLock     *sync.Mutex
+	syncTime     *time.Time
+}
+
+func createObjInfo(un *unstructured.Unstructured) *node {
+	ownerRefs := un.GetOwnerReferences()
+	// Special case for endpoint. Remove after https://github.com/kubernetes/kubernetes/issues/28483 is fixed
+	if un.GetKind() == kube.EndpointsKind && len(un.GetOwnerReferences()) == 0 {
+		ownerRefs = append(ownerRefs, metav1.OwnerReference{
+			Name:       un.GetName(),
+			Kind:       kube.ServiceKind,
+			APIVersion: "",
+		})
+	}
+	info := &node{
+		resourceVersion: un.GetResourceVersion(),
+		ref: v1.ObjectReference{
+			APIVersion: un.GetAPIVersion(),
+			Kind:       un.GetKind(),
+			Name:       un.GetName(),
+			Namespace:  un.GetNamespace(),
+		},
+		ownerRefs: ownerRefs,
+		parents:   make(map[string]*node),
+		children:  make(map[string]*node),
+		tags:      getTags(un),
+	}
+	if labels := un.GetLabels(); len(ownerRefs) == 0 && labels != nil && labels[common.LabelApplicationName] != "" {
+		info.appName = labels[common.LabelApplicationName]
+		info.resource = un
+	}
+	return info
+}
+
+func (c *clusterInfo) synced() bool {
+	return c.syncTime != nil && time.Now().Before(c.syncTime.Add(clusterSyncTimeout))
+}
+
+func (c *clusterInfo) ensureSynced() error {
+	if c.synced() {
+		return nil
+	}
+	c.syncLock.Lock()
+	defer c.syncLock.Unlock()
+	if c.synced() {
+		return nil
+	}
+	log.Infof("Start syncing cluster %s", c.cluster.Server)
+	c.nodes = make(map[string]*node)
```

Comment:
Instead of updating `c.nodes` in-place, I think you should create a temporary node map and reassign `c.nodes` at the end of this method. This way, you may not even have to hold the syncLock and will update the map atomically.

Reply:
The syncLock is required to modify `nodes` map and to prepare cluster for processing events. The cluster should be synced before processing events from watch, so we have to keep using syncLock .